cowbird/handlers/impl/magpie.py

```diff
@@ -19,6 +19,16 @@
 
 COOKIES_TIMEOUT = 60
 
+WFS_READ_PERMISSIONS = ["describefeaturetype", "describestoredqueries", "getcapabilities", "getfeature", "getgmlobject",
+                        "getpropertyvalue", "liststoredqueries"]
+WFS_WRITE_PERMISSIONS = ["createstoredquery", "dropstoredquery", "getfeaturewithlock", "lockfeature", "transaction"]
+WMS_READ_PERMISSIONS = ["describelayer", "getcapabilities", "getfeatureinfo", "getlegendgraphic", "getmap"]
+WPS_READ_PERMISSIONS = ["describeprocess", "getcapabilities"]
+WPS_WRITE_PERMISSIONS = ["execute"]
+
+LAYER_READ_PERMISSIONS = WFS_READ_PERMISSIONS + WMS_READ_PERMISSIONS
+LAYER_WRITE_PERMISSIONS = WFS_WRITE_PERMISSIONS
```

Comment:
I am not sure about the `Permission` class, since in Cowbird, it combines resource and permission info, all received from a webhook.

I agree with you though for all the name of the permissions (`describefeaturetype`, `describestoredqueries`, etc.) as they shouldn't be duplicated here ideally.

I could check to get those thos values directly from the [Permission](https://github.com/Ouranosinc/Magpie/blob/master/magpie/permissions.py#L32) enum from Magpie for example. Same thing for permission access and scopes.

I am wondering though, I don't think those permissions are classified as it is here (per WFS/WMS/WPS and READ/WRITE types for example. Should those categories be transfered to Magpie?

Reply:
Just to be sure, I would add `git+https://github.com/Ouranosinc/Magpie` to the requirements , to use Magpie. We don't have a Magpie package elsewhere than the actual github link, right?

Reply:
You're right. For the permission, maybe `PermissionSet` is more appropriate. If it is too much refactor for that case, its ok to keep Cowbird's `Permission`. For the permissions per Service/Resource, you need to use the models in https://github.com/Ouranosinc/Magpie/blob/master/magpie/services.py. With the relevant service implementations, you will have utilities like `ServiceInterface.nested_resource_allowed`, `ServiceInterface.get_resource_permissions` and so on that will resolve all the valid combinations. The logic will be 1:1 correct mapping.

And yes, you need to install directly from the repository for now.

Reply:
I can get the allowed permissions on a Layer or workspace type resource, but I don't think there is a way to know from Magpie if it a read or write type permission, is it?

So, I could keep the definitions above (`WFS_READ_PERMISSIONS`, `LAYER_READ_PERMISSIONS`, and others) and use the permission names from Magpie instead of the hardcoded values?
Else I think we would need to update Magpie with the read and write definitions, but I don't know if it is outside the scope of this PR.

Reply:
Shouldn't the mapping actually be defined by the cowbird config, as in https://github.com/Ouranosinc/cowbird/blob/master/config/config.example.yml#L113 ?

Reply:
I think this might be good for the `permission_created` case, where a single Magpie permission was created, and we send the event to cowbird. Cowbird would update the file permissions for Geoserver, and it would directly ask Magpie to update all the other `read` or `write` permissions using the permissions_synchroniser config. This seem better than relying on the `chmod` triggering an `on_modified` event on cowbird which would then update all other magpie permissions.

But the definitions above could still be useful for the `on_created` and `on_modified` case. I think it makes more sense to send all `read` or `write` permissions in one go after a file was created on Geoserver. 
We could bypass this by sending an arbitrary `read` or `write` permission to Magpie and using the config to synchronize the rest of the `read`/`write` permissions after that, but it would make less sense doing it that way.

Reply:
Maybe the correct approach would be to have a configurable input in the `GeoServer` handler definition that defines permission mapping? 
Or maybe extend `sync_permissions` in a way that doesn't assume Magpie-only handler?
Would it be possible to pass around a "source" handler and a "target" handler, the default being (Magpie, Magpie) to support the current definition?